testsuite/model/src/test/java/org/keycloak/testsuite/model/session/OfflineSessionPersistenceTest.java

```diff
@@ -257,8 +257,8 @@ public void testPersistenceMultipleNodesClientSessionsAtRandomNode() throws Inte
                     // IllegalLifecycleStateException: ISPN000324: Cache 'clientSessions' is in 'STOPPING' state and this is an invocation not belonging to an
                     // on-going transaction, so it does not accept new invocations."
                     // also: org.infinispan.commons.CacheException: java.lang.IllegalStateException: Read commands must ignore leavers
-                    if ((ex.getCause() != null && ex.getCause().getMessage().contains("ISPN000324") ||
-                            (ex instanceof CacheException && ex.getMessage().contains("Read commands must ignore leavers")))) {
+                    if (ex.getCause() != null && (ex.getCause().getMessage().contains("ISPN000324") || ex.getCause().getMessage().contains("ISPN000217")) ||
```

Comment:
After looking at the stack trace (see below), I suspect that the ISPN000217 is not in the cause, but the parent message. This then also requires re-arranging the parentheses again. See below for a suggested change.

There is also a null check, as not every exception has a message.

```suggestion
                    if ((ex.getCause() != null && ex.getCause().getMessage().contains("ISPN000324")) || (ex.getMessage() != null && ex.getMessage().contains("ISPN000217")) ||
```

```
2022-09-24T00:57:52.4568276Z 00:57:52,387 ERROR [org.keycloak.testsuite.model.KeycloakModelParameters] (Time-limited test) Execution didn't complete
2022-09-24T00:57:52.4569944Z java.util.concurrent.ExecutionException: java.lang.RuntimeException: org.infinispan.remoting.RemoteException: ISPN000217: Received exception from node-8(site-id=site-2), see cause for remote stack trace
2022-09-24T00:57:52.4570866Z 	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
2022-09-24T00:57:52.4571307Z 	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:205)
2022-09-24T00:57:52.4571869Z 	at org.keycloak.testsuite.model.KeycloakModelTest.inIndependentFactories(KeycloakModelTest.java:395)
2022-09-24T00:57:52.4573167Z 	at org.keycloak.testsuite.model.session.OfflineSessionPersistenceTest.testPersistenceMultipleNodesClientSessionsAtRandomNode(OfflineSessionPersistenceTest.java:246)
2022-09-24T00:57:52.4574150Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2022-09-24T00:57:52.4574846Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2022-09-24T00:57:52.4575490Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2022-09-24T00:57:52.4576014Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2022-09-24T00:57:52.4576462Z 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
2022-09-24T00:57:52.4576995Z 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2022-09-24T00:57:52.4577544Z 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
2022-09-24T00:57:52.4578104Z 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2022-09-24T00:57:52.4578682Z 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
2022-09-24T00:57:52.4579270Z 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
2022-09-24T00:57:52.4579770Z 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2022-09-24T00:57:52.4580140Z 	at java.base/java.lang.Thread.run(Thread.java:829)
2022-09-24T00:57:52.4580868Z Caused by: java.lang.RuntimeException: org.infinispan.remoting.RemoteException: ISPN000217: Received exception from node-8(site-id=site-2), see cause for remote stack trace
2022-09-24T00:57:52.4581553Z 	at org.keycloak.testsuite.model.KeycloakModelTest.inIndependentFactory(KeycloakModelTest.java:466)
2022-09-24T00:57:52.4582204Z 	at org.keycloak.testsuite.model.KeycloakModelTest.lambda$inIndependentFactories$5(KeycloakModelTest.java:366)
2022-09-24T00:57:52.4582727Z 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2022-09-24T00:57:52.4583212Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2022-09-24T00:57:52.4583752Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2022-09-24T00:57:52.4584091Z 	... 1 more
2022-09-24T00:57:52.4584668Z Caused by: org.infinispan.remoting.RemoteException: ISPN000217: Received exception from node-8(site-id=site-2), see cause for remote stack trace
2022-09-24T00:57:52.4585330Z 	at org.infinispan.remoting.transport.ResponseCollectors.wrapRemoteException(ResponseCollectors.java:25)
2022-09-24T00:57:52.4585961Z 	at org.infinispan.commands.write.ExceptionAckCommand.ack(ExceptionAckCommand.java:39)
2022-09-24T00:57:52.4586820Z 	at org.infinispan.remoting.inboundhandler.TrianglePerCacheInboundInvocationHandler.handleBackupAckCommand(TrianglePerCacheInboundInvocationHandler.java:183)
2022-09-24T00:57:52.4587891Z 	at org.infinispan.remoting.inboundhandler.TrianglePerCacheInboundInvocationHandler.handle(TrianglePerCacheInboundInvocationHandler.java:98)
2022-09-24T00:57:52.4588856Z 	at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler.handleCacheRpcCommand(GlobalInboundInvocationHandler.java:165)
2022-09-24T00:57:52.4589794Z 	at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler.handleFromCluster(GlobalInboundInvocationHandler.java:112)
2022-09-24T00:57:52.4590604Z 	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processRequest(JGroupsTransport.java:1472)
2022-09-24T00:57:52.4591410Z 	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processMessage(JGroupsTransport.java:1395)
2022-09-24T00:57:52.4592083Z 	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.access$300(JGroupsTransport.java:146)
2022-09-24T00:57:52.4592729Z 	at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.up(JGroupsTransport.java:1586)
2022-09-24T00:57:52.4593217Z 	at org.jgroups.JChannel.up(JChannel.java:780)
2022-09-24T00:57:52.4593597Z 	at org.jgroups.stack.ProtocolStack.up(ProtocolStack.java:913)
2022-09-24T00:57:52.4593960Z 	at org.jgroups.protocols.FRAG3.up(FRAG3.java:165)
2022-09-24T00:57:52.4594340Z 	at org.jgroups.protocols.FlowControl.up(FlowControl.java:355)
2022-09-24T00:57:52.4594802Z 	at org.jgroups.protocols.FlowControl.up(FlowControl.java:363)
2022-09-24T00:57:52.4595198Z 	at org.jgroups.protocols.pbcast.GMS.up(GMS.java:876)
2022-09-24T00:57:52.4595596Z 	at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:254)
2022-09-24T00:57:52.4596032Z 	at org.jgroups.protocols.UNICAST3.deliverMessage(UNICAST3.java:1048)
2022-09-24T00:57:52.4596470Z 	at org.jgroups.protocols.UNICAST3.addMessage(UNICAST3.java:771)
2022-09-24T00:57:52.4596917Z 	at org.jgroups.protocols.UNICAST3.handleDataReceived(UNICAST3.java:752)
2022-09-24T00:57:52.4597352Z 	at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:405)
2022-09-24T00:57:52.4597731Z 	at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:592)
2022-09-24T00:57:52.4598139Z 	at org.jgroups.protocols.VERIFY_SUSPECT.up(VERIFY_SUSPECT.java:132)
2022-09-24T00:57:52.4598578Z 	at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:186)
2022-09-24T00:57:52.4598981Z 	at org.jgroups.protocols.FD_SOCK.up(FD_SOCK.java:254)
2022-09-24T00:57:52.4599332Z 	at org.jgroups.protocols.MERGE3.up(MERGE3.java:281)
2022-09-24T00:57:52.4599705Z 	at org.jgroups.protocols.Discovery.up(Discovery.java:300)
2022-09-24T00:57:52.4600090Z 	at org.jgroups.protocols.TP.passMessageUp(TP.java:1400)
2022-09-24T00:57:52.4600563Z 	at org.jgroups.util.SubmitToThreadPool$SingleMessageHandler.run(SubmitToThreadPool.java:98)
2022-09-24T00:57:52.4600919Z 	... 3 more
2022-09-24T00:57:52.4601238Z 	Suppressed: org.infinispan.commons.util.logging.TraceException
2022-09-24T00:57:52.4601804Z 		at org.infinispan.interceptors.impl.SimpleAsyncInvocationStage.get(SimpleAsyncInvocationStage.java:39)
2022-09-24T00:57:52.4602492Z 		at org.infinispan.interceptors.impl.AsyncInterceptorChainImpl.invoke(AsyncInterceptorChainImpl.java:249)
2022-09-24T00:57:52.4603107Z 		at org.infinispan.cache.impl.InvocationHelper.doInvoke(InvocationHelper.java:297)
2022-09-24T00:57:52.4603629Z 		at org.infinispan.cache.impl.InvocationHelper.invoke(InvocationHelper.java:101)
2022-09-24T00:57:52.4604287Z 		at org.infinispan.cache.impl.InvocationHelper.invoke(InvocationHelper.java:83)
2022-09-24T00:57:52.4604791Z 		at org.infinispan.cache.impl.CacheImpl.putIfAbsent(CacheImpl.java:1333)
2022-09-24T00:57:52.4605295Z 		at org.infinispan.cache.impl.DecoratedCache.putIfAbsent(DecoratedCache.java:688)
2022-09-24T00:57:52.4605821Z 		at org.infinispan.cache.impl.DecoratedCache.putIfAbsent(DecoratedCache.java:277)
2022-09-24T00:57:52.4606403Z 		at org.infinispan.cache.impl.AbstractDelegatingCache.putIfAbsent(AbstractDelegatingCache.java:151)
2022-09-24T00:57:52.4606978Z 		at org.infinispan.cache.impl.EncoderCache.putIfAbsent(EncoderCache.java:238)
2022-09-24T00:57:52.4607790Z 		at org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction.runOperationInCluster(InfinispanChangelogBasedTransaction.java:203)
2022-09-24T00:57:52.4608859Z 		at org.keycloak.models.sessions.infinispan.changes.InfinispanChangelogBasedTransaction.commitImpl(InfinispanChangelogBasedTransaction.java:172)
2022-09-24T00:57:52.4609662Z 		at org.keycloak.models.AbstractKeycloakTransaction.commit(AbstractKeycloakTransaction.java:48)
2022-09-24T00:57:52.4610323Z 		at org.keycloak.services.DefaultKeycloakTransactionManager.commit(DefaultKeycloakTransactionManager.java:146)
2022-09-24T00:57:52.4611116Z 		at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:256)
2022-09-24T00:57:52.4611768Z 		at org.keycloak.testsuite.model.KeycloakModelTest.inComittedTransaction(KeycloakModelTest.java:543)
2022-09-24T00:57:52.4612434Z 		at org.keycloak.testsuite.model.KeycloakModelTest.inComittedTransaction(KeycloakModelTest.java:538)
2022-09-24T00:57:52.4613041Z 		at org.keycloak.testsuite.model.KeycloakModelTest.withRealm(KeycloakModelTest.java:566)
2022-09-24T00:57:52.4613820Z 		at org.keycloak.testsuite.model.session.OfflineSessionPersistenceTest.createOfflineClientSession(OfflineSessionPersistenceTest.java:360)
2022-09-24T00:57:52.4614913Z 		at org.keycloak.testsuite.model.session.OfflineSessionPersistenceTest.lambda$testPersistenceMultipleNodesClientSessionsAtRandomNode$27(OfflineSessionPersistenceTest.java:253)
2022-09-24T00:57:52.4615761Z 		at org.keycloak.testsuite.model.KeycloakModelTest.lambda$inIndependentFactories$4(KeycloakModelTest.java:373)
2022-09-24T00:57:52.4616410Z 		at org.keycloak.testsuite.model.KeycloakModelTest.inIndependentFactory(KeycloakModelTest.java:464)
2022-09-24T00:57:52.4617060Z 		at org.keycloak.testsuite.model.KeycloakModelTest.lambda$inIndependentFactories$5(KeycloakModelTest.java:366)
2022-09-24T00:57:52.4617579Z 		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2022-09-24T00:57:52.4617874Z 		... 3 more
2022-09-24T00:57:52.4618307Z Caused by: org.infinispan.commons.IllegalLifecycleStateException: Cannot wire or start components while the registry is not running
2022-09-24T00:57:52.4618941Z 	at org.infinispan.marshall.exts.ThrowableExternalizer.readObject(ThrowableExternalizer.java:215)
2022-09-24T00:57:52.4619565Z 	at org.infinispan.marshall.exts.ThrowableExternalizer.readObject(ThrowableExternalizer.java:42)
2022-09-24T00:57:52.4620194Z 	at org.infinispan.marshall.core.GlobalMarshaller.readWithExternalizer(GlobalMarshaller.java:727)
2022-09-24T00:57:52.4620836Z 	at org.infinispan.marshall.core.GlobalMarshaller.readNonNullableObject(GlobalMarshaller.java:708)
2022-09-24T00:57:52.4621483Z 	at org.infinispan.marshall.core.GlobalMarshaller.readNullableObject(GlobalMarshaller.java:357)
2022-09-24T00:57:52.4622074Z 	at org.infinispan.marshall.core.BytesObjectInput.readObject(BytesObjectInput.java:32)
2022-09-24T00:57:52.4622651Z 	at org.infinispan.commands.write.ExceptionAckCommand.readFrom(ExceptionAckCommand.java:58)
2022-09-24T00:57:52.4623316Z 	at org.infinispan.marshall.exts.TriangleAckExternalizer.exceptionAckCommand(TriangleAckExternalizer.java:63)
2022-09-24T00:57:52.4624006Z 	at org.infinispan.marshall.exts.TriangleAckExternalizer.readObject(TriangleAckExternalizer.java:46)
2022-09-24T00:57:52.4624654Z 	at org.infinispan.marshall.exts.TriangleAckExternalizer.readObject(TriangleAckExternalizer.java:24)
2022-09-24T00:57:52.4625283Z 	at org.infinispan.marshall.core.GlobalMarshaller.readWithExternalizer(GlobalMarshaller.java:727)
2022-09-24T00:57:52.4625927Z 	at org.infinispan.marshall.core.GlobalMarshaller.readNonNullableObject(GlobalMarshaller.java:708)
2022-09-24T00:57:52.4626569Z 	at org.infinispan.marshall.core.GlobalMarshaller.readNullableObject(GlobalMarshaller.java:357)
2022-09-24T00:57:52.4627204Z 	at org.infinispan.marshall.core.GlobalMarshaller.objectFromObjectInput(GlobalMarshaller.java:191)
2022-09-24T00:57:52.4627845Z 	at org.infinispan.marshall.core.GlobalMarshaller.objectFromByteBuffer(GlobalMarshaller.java:220)
2022-09-24T00:57:52.4628524Z 	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processRequest(JGroupsTransport.java:1455)
2022-09-24T00:57:52.4628975Z 	... 25 more
```

Reply:
Your analysis seems correct. I updated the PR. Thanks.